src/common/input/InputManager.h

```diff
@@ -92,44 +103,44 @@ struct SBCOutput {
 
 #pragma pack()
 
+union InputBuff {
+	XpadInput ctrl;
+	SBCInput sbc;
+};
 
-// hle specific input types
-typedef struct _CXBX_XINPUT_DEVICE_INFO {
-	uint8_t  ucType;            // xbox controller type
-	uint8_t  ucSubType;         // xbox controller subtype
-	uint8_t  ucInputStateSize;  // xbox controller input state size in bytes, not include dwPacketNumber
-	uint8_t  ucFeedbackSize;    // xbox controller feedback size in bytes, not include FeedbackHeader
+struct DeviceInfo {
+	xbox::HANDLE hhandle;      // device handle returned by xapi
+	bool bAutoPoll;            // autopoll on/off, as instructed by the title in XInputOpen
+	bool bAutoPollDefault;     // default autopoll value, depending on device type
+	uint8_t ucType;            // xapi type
+	uint8_t ucSubType;         // xapi subtype
+	uint8_t ucInputStateSize;  // input state size in bytes, does not include dwPacketNumber
+	uint8_t ucFeedbackSize;    // feedback size in bytes, does not include FeedbackHeader
 	uint32_t dwPacketNumber;
-}
-CXBX_XINPUT_DEVICE_INFO, *PCXBX_XINPUT_DEVICE_INFO;
+	InputBuff buff;
+};
 
-union CXBX_XINPUT_IN_STATE {
-	XpadInput Gamepad;
-	SBCInput SBC;
+struct DeviceState {
+	std::string port;
+	int port_idx;
+	XBOX_INPUT_DEVICE type;
+	bool bPendingRemoval;
+	bool bSignaled;
+	DeviceInfo info;
+	DeviceState *slots[XBOX_CTRL_NUM_SLOTS];
+	DeviceState *upstream;
 };
 
-// this structure is for use of tracking the xbox controllers assigned to 4 ports.
-typedef struct _CXBX_CONTROLLER_HOST_BRIDGE {
-	HANDLE                  hXboxDevice;      // xbox device handle to this device, we use the address of this bridge as the handle, only set after opened. cleared after closed.
-	int                     XboxPort;         // xbox port# for this xbox controller
-	XBOX_INPUT_DEVICE       XboxType;         // xbox device type
-	CXBX_XINPUT_IN_STATE    *InState;
-	bool                    bPendingRemoval;
-	bool                    bSignaled;
-	bool                    bIoInProgress;
-	bool                    bAutoPoll;         // autopoll on/off, as instructed by the title in XInputOpen
-	bool                    bAutoPollDefault;  // default autopoll value, depending on device type
-	CXBX_XINPUT_DEVICE_INFO XboxDeviceInfo;
-}
-CXBX_CONTROLLER_HOST_BRIDGE, *PCXBX_CONTROLLER_HOST_BRIDGE;
+extern DeviceState g_devs[4 + 8];
```

Comment:
```suggestion
extern DeviceState g_devs[4 + 4 * XBOX_CTRL_NUM_SLOTS];
```
imo, ^ will better than hardcode numbers.
Although, it appear we don't have macro for `XBOX_HW_NUM_PORT_SLOTS` or the like.